datasource/src/main/scala/quasar/physical/mongo/interpreter/Cartesian.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo.interpreter
+
+import slamdata.Predef._
+
+import cats.syntax.traverse._
+import cats.instances.option._
+import cats.instances.list._
+
+import quasar.common.CPathField
+import quasar.physical.mongo.MongoExpression
+import quasar.physical.mongo.{Interpretation, Interpreter, Aggregator, Version, MongoExpression => E}
+import quasar.ScalarStage
+
+import shims._
+
+object Cartesian {
+  def apply(
+      uniqueKey: String,
+      version: Version,
+      cartouches: Map[CPathField, (CPathField, List[ScalarStage.Focused])],
+      interpreter: Interpreter)
+      : Option[List[Aggregator]] = {
+
+    if (cartouches.isEmpty) Some(List(Aggregator.filter(E.Object(uniqueKey -> E.Bool(false)))))
+    else {
+      val cartoucheList = cartouches.toList
+
+      val interpretations: Option[List[List[Aggregator]]] =
+        cartoucheList.traverse {
+          case (alias, (_, instructions)) =>
+            val interpreted: Interpretation =
+              interpreter.refineInterpretation(alias.name, Interpretation.initial(instructions))
+            if (!interpreted.stages.isEmpty) None
+            else Some(interpreted.aggregators)
+        }
+
+      interpretations map { is =>
+        val defaultPairs: List[(String, MongoExpression)] =
+          cartoucheList map {
+            case (alias, _) => (alias.name, E.key(alias.name))
+          }
+        val defaultObject = E.Object(defaultPairs:_*)
+
+        val initialProjectionPairs = cartoucheList map {
+          case (alias, (field, instructions)) => alias.name -> (E.key(uniqueKey) +/ E.key(field.name))
+        }
+
+        val initialProjection =
+          Aggregator.project(E.Object(initialProjectionPairs:_*))
+
+        val lastProjectionPairs = cartoucheList map {
+          case (alias, _) => alias.name -> E.key(alias.name)
+        }
+        val lastProjection = Aggregator.project(E.Object(uniqueKey -> E.Object(lastProjectionPairs:_*)))
+
+        val instructions = is flatMap (_ flatMap {
+          case Aggregator.project(a @ E.Object(_*)) =>
+            List(Aggregator.project(defaultObject + a))
```

Comment:
What do you mean? The sequence of action is following 

Having `Cartesian (Map { alias, (name, instruction) })`

+ Having cartesian for every cartouche we create a projection from projected key to alias, then we have `{alias: $name}`
+ Then we interpret all instrucitons on `alias` field
+ Then we remove `nonNull` aggregator and from this instructions
+ Then we make object `uniqueKey: { alias: interpretationResult}` 
+ Then we add `nonNull` back

Reply:
This appears to be modifying all cartouche projections, merging them with `defaultObject`, I was just confirming that this was correct for the projection in a cartouche like `Mask(. -> Object), Pivot(ExcludeId, Object), Project(.x.y.z)` since the project is not being applied to the input of the cartesian.

Reply:
Having `alias -> key -> Mask(. -> Object), Pivot(whatever, Object), Project(.x.y.z) ;; alias2 -> whatever` all constructed `$project`s need to have both `alias` and `alias2` fields, but most of interpretations constructs `$project` with only one key. So `$project` constructed for `alias` modified to have both keys by merging with default object, that is simply preservationi of all aliases in cartesian, i.e. in this example it would be `alias2 -> $alias2` for projects working with `alias` and vice versa `alias -> $alias` for projects working with `alias2`

Reply:
I see, thanks for explaining. :+1: